Snakemake/rules/merge_samples_vcf.smk

```diff
@@ -4,113 +4,108 @@
 #
 import itertools
 import numpy as np
-configfile: 'test.yaml'
+configfile: 'config.yaml'
 
-def list_full_paths(directory):
-    return [os.path.join(directory, file) for file in os.listdir(directory)]
+dir = config['vcfDir']
+# single file patterns
+(PREFIX, SUFFIX) = glob_wildcards(dir+'/{prefix}_{suffix}.vcf')
+#print(PREFIX)
 
-def dic_files_per_chromosome(direc_lst, noChromosomes):
-    file_dic={}
-    for i in range(1, noChromosomes + 1):
-        file_dic[i] = [name for name in direc_lst if str(i) in name]
-
-    return file_dic
 
 rule all:
     input:
-        expand('Chr{chromosome}.vcf.gz', chromosome=range(1, config['noChromosomes'] + 1))
-
-#workdir: config['vcfDir']
+        expand('{prefix}_{suffix}', zip, prefix=(PREFIX), suffix=SUFFIX)
+        #expand('Chr{chromosome}.vcf.gz', chromosome=range(1, config['noChromosomes'] + 1))
 
-VCFs =[x for x in list_full_paths(config['vcfDir']) if x.endswith(".vcf")]
-print('here are vcfs', VCFs)
-
-file_dic = dic_files_per_chromosome(VCFs, config['noChromosomes'])
-count = list(set(map(len, file_dic.values())))[0]
-#print(file_dic)
-#print(count)
 
 rule extract_samples_list:
-    input:
-        files = lambda wildcards: expand(
-            ['{files}'], files = file_dic[int(wildcards.chromosome)]
-        )
-    output:
-        expand('ids{{chromosome}}_{i}', i = range(1, count + 1))
+    input: dir + '/{prefix}_{suffix}.vcf'
+    output: '{prefix}_{suffix}.txt'
     shell:
         """
-        vcfs=( {input.files} )
-        outfiles=( {output} )
-
-        for ((i=0; i<${{#vcfs[@]}}; i++)); do
-            bcftools query -l ${{vcfs[$i]}} > ${{outfiles[$i]}}
-        done
+        bcftools query -l {input} > {output}
         """
 
+#rule extract_samples_list:
+#    input:
+#        expand(dir + '/{prefix}_{suffix}.vcf', zip, prefix=PREFIX, suffix=SUFFIX)
+#        files = lambda wildcards: expand(
+#            ['{files}'], files = file_dic[int(wildcards.chromosome)]
+#        )
+#    output:
+#        expand('{prefix}_{suffix}.txt', zip, prefix=PREFIX, suffix=SUFFIX)
+        #expand('ids{{chromosome}}_{i}', i = range(1, count + 1))
+#    shell:
+#        """
+#        vcfs=( {input} )
+#        outfiles=( {output} )
+
+#        for ((i=0; i<${{#vcfs[@]}}; i++)); do
+#            bcftools query -l ${{vcfs[$i]}} > ${{outfiles[$i]}}
+#        done
+
+#        """
+
 rule compare_samples:
     input:
-        rules.extract_samples_list.output
+#    # takes input from previous rule
+        expand('{prefix}_{suffix}.txt', zip, prefix=PREFIX, suffix=SUFFIX)
```

Comment:
I don't think you need this expand here - if the number of input and output files is the same (they link), they you don't need expand. You would need it when you have multiple input files and only one output (to sort of merge all inputs into one output).

Reply:
Ah, gotcha ... That's tedious